viabel/optimization.py

```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
+        K_max : `int`
+            Number of iterations of the optimization
+        objective: `function`
+            Function for constructing the objective and gradient function
+        init_param : `numpy.ndarray`, shape(var_param_dim,)
+            Initial values of the variational parameters
+
+        """
+        if not objective.approx.supports_kl:
+            print('WARNING: approximation family does not support KL. Using FASO.',
+                  flush=True)
+            return super().optimize(K_max, objective, init_param)
+        k_new = -1 #Number of iterations at the given learning rate
+        k = 0 #Number of times learning rate decreases
+        k_total = 0 #Total number of iterations
+        k_add = 0 #Iteration number when convergence reached for a fixed step size
+        k_stopped_final = None #Iteration number when stopping rule criteria met
+        sgo = self._sgo
+        diagnostics = self._sgo._diagnostics
+        if isinstance(self._sgo, AveragedRMSProp) or isinstance(self._sgo, AveragedAdam):
+            reg_model = StanModel_cache(model_name='weighted_lin_regression_sgd')
+        else:
+            reg_model = StanModel_cache(model_name='weighted_lin_regression')
+        iterate_average_curr = init_param.copy()
+        iterate_average_curr_hist = [iterate_average_curr]
+        variational_param_history = None
+        value_history = None
+        grad_history = None
+        descent_dir_history = None
+        iterate_average_k_history = None
+        iterate_average_history = None
+        ess_and_mcse_k_history = None
+        ess_history = None
+        mcse_history = None
+        final_mcse_history = []
+        SKL_history = []
+        learning_rate_hist = []
+        kappa_hist = []
+        c_hist = []
+        c_sample_hist = []
+        kappa_sample_hist = []
+        k_new_hist = [0]
+        k_conv_hist = []
+        k_Rhat_hist = []
+        stopped = False
+        convg_iters_hist = []
+        predicted_iters_hist = []
+        stopping_crt = []
+        k_stopped_final_hist = []
+        LR = [] #learning rate for WLS regression 
+        CI = [] #Converged Iterations for WLS regression
+        try:
+            while not stopped:
+                K_max -= (k_new +1)
+                iterate_average_prev = iterate_average_curr
+                if k==0 and self._init_rmsprop:
+                    rmsprop = RMSProp(learning_rate=sgo._learning_rate, diagnostics=diagnostics)
+                    faso = FASO(sgo = rmsprop)
+                    opt = faso.optimize(K_max, objective, iterate_average_curr)
+                else:
+                    opt = super().optimize(K_max, objective, iterate_average_curr)
+                if opt['k_stopped'] is not None and k!=0: #removing the number of convergence iterations of initial learning rate
+                    convg_iters = opt['k_stopped']
+                    convg_iters_hist.append(convg_iters)
+                    CI.append(convg_iters)
```

Comment:
same question I had about `LR`. Why the two? Can we avoid the (near?) duplications?